From 67bd3d73e3f9dea03e5e5269dbfc2ea3a15db1cd Mon Sep 17 00:00:00 2001
From: Russell King <rmk+kernel@arm.linux.org.uk>
Date: Thu, 30 Oct 2014 15:53:14 +0000
Subject: [PATCH 063/149] staging: etnaviv: move mutex around
 component_{un,}bind_all()

Hold the mutex while calling component_{un,}bind_all() so that
components can perform initialisation in their bind and unbind
callbacks from the component helper.

Signed-off-by: Russell King <rmk+kernel@arm.linux.org.uk>
---
 drivers/staging/etnaviv/etnaviv_drv.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/drivers/staging/etnaviv/etnaviv_drv.c b/drivers/staging/etnaviv/etnaviv_drv.c
index 80b3c66..158cb38 100644
--- a/drivers/staging/etnaviv/etnaviv_drv.c
+++ b/drivers/staging/etnaviv/etnaviv_drv.c
@@ -99,10 +99,11 @@ static int etnaviv_unload(struct drm_device *dev)
 		if (g)
 			etnaviv_gpu_pm_suspend(g);
 	}
-	mutex_unlock(&dev->struct_mutex);
 
 	component_unbind_all(dev->dev, dev);
 
+	mutex_unlock(&dev->struct_mutex);
+
 	dev->dev_private = NULL;
 
 	kfree(priv);
@@ -116,8 +117,6 @@ static void load_gpu(struct drm_device *dev)
 	struct etnaviv_drm_private *priv = dev->dev_private;
 	unsigned int i;
 
-	mutex_lock(&dev->struct_mutex);
-
 	for (i = 0; i < ETNA_MAX_PIPES; i++) {
 		struct etnaviv_gpu *g = priv->gpu[i];
 
@@ -132,8 +131,6 @@ static void load_gpu(struct drm_device *dev)
 			}
 		}
 	}
-
-	mutex_unlock(&dev->struct_mutex);
 }
 
 static int etnaviv_load(struct drm_device *dev, unsigned long flags)
@@ -157,12 +154,16 @@ static int etnaviv_load(struct drm_device *dev, unsigned long flags)
 
 	platform_set_drvdata(pdev, dev);
 
+	mutex_lock(&dev->struct_mutex);
+
 	err = component_bind_all(dev->dev, dev);
 	if (err < 0)
 		return err;
 
 	load_gpu(dev);
 
+	mutex_unlock(&dev->struct_mutex);
+
 	return 0;
 }
 
-- 
2.6.2

